README.md
# fl_animated_linechart

[![Codemagic build status](https://api.codemagic.io/apps/5d5e513ff8278e001ca52adf/5d5e513ff8278e001ca52ade/status_badge.svg)](https://codemagic.io/apps/5d5e513ff8278e001ca52adf/5d5e513ff8278e001ca52ade/latest_build)

![Animations](chart.gif)

An animated chart library for flutter.
 - Support for datetime axis
 - Multiple y axis, supporting different units
 - Highlight selection
 - Animation of the chart
 - Possibility of adding legends
 - Support for both horizontal and vertical markerlines 
 - Tested with more than 3000 points and still performing

There are currently two different charts:
 - line chart
 - area chart

## Getting Started

Try the sample project or include in your project.

Highlight for the line chart:

![Chart example with highlight](withSelection.png)

Linechart with markerlines, icons, legends and "heat-map" shaded zones:

![Chart example with markerlines and legends](withMarkerlinesAndLegends.png)

![Chart example with horizontal and vertical markerlines along with icons on the chart and legends](withHorizontalAndVerticalMarkerlinesAndLegends.png)

![Chart example with shaded area between markerlines](withShadedAreaBetweenMarkerlines.png)

Area chart:

![Area Chart example](areaChart.png)



![Area Chart example](areaChartGradient.png)

Example code:
```dart
    LineChart lineChart = LineChart.fromDateTimeMaps([line1, line2], [Colors.green, Colors.blue]);

    return Scaffold(
      appBar: AppBar(
        title: Text(widget.title),
      ),
      body: Container(
        child: Column(
            mainAxisSize: MainAxisSize.max,
            mainAxisAlignment: MainAxisAlignment.spaceBetween,
            crossAxisAlignment: CrossAxisAlignment.stretch,
            children: [
              Expanded(child: AnimatedLineChart(lineChart)),
            ]
        ),
      ),
    );
```
<br/>
Example code with horizontal markerlines and legends:

    LineChart lineChart = LineChart.fromDateTimeMaps([
        line1,
        line2,
        line3,
        line4,
        line5
      ], [
        Colors.blue,
        Colors.red,
        Colors.yellow,
        Colors.yellow,
        Colors.red
      ], [
        'C',
        'C',
        'C',
        'C',
        'C',
      ], tapTextFontWeight: FontWeight.w400);

  To define a line as a horizontal dashed markerline, first define the line as any other regular line, then set isMarkerLine to true:<br/>

    lineChart.lines[1].isMarkerLine = true;
    lineChart.lines[2].isMarkerLine = true;
    lineChart.lines[3].isMarkerLine = true;
    lineChart.lines[4].isMarkerLine = true;
      
  Or: <br/>
  ```dart
    lineChart.lines.skip(1).forEach((line) {
        line.isMarkerLine = true;
      });
  ```
  Legends: <br/>
  A Legend has the following constructor: <br/>
  ```dart
      const Legend({this.title, this.color, this.icon, this.style, this.showLeadingLine = false,});
  ```
    
  You can choose to either have a short line or an Icon as the first part of the legend before the title. It is possible set showLeadingLine to true, then a short line will be shown in the color you have defined. <br/>
  You can change the Color of the line and Icon, and the TextStyle of the String title. <br/>
  The length of the Legends list must to equal to the length of chartlines which have isMakerLine = true. <br/>
  
  When orientation of the screen is in landscape mode, it is possible to have the legends drawn on the right hand side instead of below the graph. 
  Simply set ``` legendsRightLandscapeMode: true ``` <br/>

```dart 
      return Scaffold(
      appBar: AppBar(
        title: Text(widget.title),
      ),
      body: Container(
        child: Column(
          mainAxisSize: MainAxisSize.max,
          mainAxisAlignment: MainAxisAlignment.spaceBetween,
          crossAxisAlignment: CrossAxisAlignment.stretch,
          children: [
            Expanded(
              child: AnimatedLineChart(
                lineChart,
                toolTipColor: Colors.white,
                gridColor: Colors.black54,
                textStyle: TextStyle(fontSize: 10, color: Colors.black54),
                showMarkerLines:
                    true, // If this value is not set to true, all defines lines will be filled lines and not dashed
                legends: [
                  Legend(title: 'Critical', color: Colors.red, showLeadingLine: true),
                  Legend(title: 'Warning', color: Colors.yellow, Icon(Icons.report_problem_rounded, size: 17, color: Colors.yellow)),
                ],
                legendsRightLandscapeMode: true,
              ),
            ),
          ],
        ),
      ),
    ); 
```
<br/>
Example code with horizontal and vertical markerlines along with icons on the chart and legends:

  It is possible to define a maximum of two vertical markerlines. <br/>
  The verticalMarker variable is a ``` List<DateTime> ``` and the length can be >= 2. <br/>
  If two vertical markerlines are defined, the area between the two lines will be filled with a color that can be defined with verticalMarkerColor.<br/>

  It is possible to add an Icon on the point where the vertical markerline crosses the y-axis value by defining the verticalMarkerIcon. <br/>
  The verticalMarkerIcon variable takes a ``` List<Icon> ``` and the lenght must be equal to the length of the verticalMarker variable. <br/>
  The possibility of adding a colored background to the icons have been added and can be defined with iconBackgroundColor which takes a Color. <br/>

```dart 
      return Scaffold(
      appBar: AppBar(
        title: Text(widget.title),
      ),
      body: Container(
        child: Column(
          mainAxisSize: MainAxisSize.max,
          mainAxisAlignment: MainAxisAlignment.spaceBetween,
          crossAxisAlignment: CrossAxisAlignment.stretch,
          children: [
            Expanded(
              child: AnimatedLineChart(
                lineChart,
                toolTipColor: Colors.white,
                gridColor: Colors.black54,
                textStyle: TextStyle(fontSize: 10, color: Colors.black54),
                showMarkerLines:
                    true, // If this value is not set to true, all defines lines will be filled lines and not dashed
                legends: [
                  Legend(title: 'Critical', color: Colors.red),
                  Legend(title: 'Warning', color: Colors.yellow),
                ],
                verticalMarker:[
                    DateTime.parse('2012-02-27 12:54:00'),
                    DateTime.parse('2012-02-27 13:16:00')
                        ],
                  verticalMarkerColor: Colors.yellow,
                  verticalMarkerIcon: [
                    Icon(
                      Icons.report_problem_rounded,
                      color: Colors.yellow,
                    ),
                    Icon(
                      Icons.check_circle_rounded,
                      color: Colors.green,
                    ),
                  ],
                  iconBackgroundColor: Colors.white,
              ),
            ),
          ],
        ),
      ),
    );
```

Example code with shaded area between markerlines:
It is possible to have shaded areas between the defined markerlines. <br/>
The shaded area will be in the same color as the markerline. <br/>
It is important that the order of enums in the ``` List<MaxMin> ``` filledMarkerLinesValues matches the order of defined markerlines to be shown in the graph. <br/>
Enums with the value MaxMin.MAX will draw to the top if there is only one markerline defined as MAX, otherwise it will draw from i - 1 where enum values are MAX. <br/>
If enum value is MaxMin.MIN the shaded area will draw downwards to i + 1, unless the index is the last in the list, then the shaded area will draw all the way to the bottom of the graph. <br/>
```dart 
      return Scaffold(
      appBar: AppBar(
        title: Text(widget.title),
      ),
      body: Container(
        child: Column(
          mainAxisSize: MainAxisSize.max,
          mainAxisAlignment: MainAxisAlignment.spaceBetween,
          crossAxisAlignment: CrossAxisAlignment.stretch,
          children: [
            Expanded(
              child: AnimatedLineChart(
                  chart,
                  key: UniqueKey(),
                  gridColor: Colors.black54,
                  textStyle: TextStyle(fontSize: 10, color: Colors.black54),
                  toolTipColor: Colors.white,
                  showMarkerLines: true,
                  fillMarkerLines: true,
                  filledMarkerLinesValues:[
                          MaxMin.MAX,
                          MaxMin.MAX,
                          MaxMin.MIN,
                          MaxMin.MIN,
                        ],
              ),
            ),
          ],
        ),
      ),
    );

```

The example app, can toggle between line chart and area chart.
![Example app](exampleScreenshot.png)


CHANGELOG.md
## [1.1.9]
* dart format files to pass static analysis

## [1.1.8]

* same as 1.1.7. Updated changelog in this version

## [1.1.7]

* Ability to set marker lines(kind of thresholds if you will), legends and heat map like shaded zones. 
* Configurable color for measurement highlighter

## [1.1.6]

* Ability to set a custom label for yAxis(still use units[index] if not provided).Thanks to @AndriiHarashchak
* Fix exception when DateTime values in series are same for all values in Map(offset in NaN exception).
* Update example app to remove deprecated FlatButton and add windows support (tested on windows 11)

## [1.1.5]

* Adding null safety. Thanks to @AndriiHarashchak.

## [1.1.4]

* Upgrade example to AndroidX and Android embedded v2

## [1.1.3]

* Fix to support any lines
* Make the pattern for date/time configurable

## [1.1.2]

* Expose parameter to configure the FontWeight of the highlighted text in charts

## [1.1.1]

* Refactor mouse over text to TapText typedef and expose it on factory method

## [1.1.0]

* Make mouse over text configurable

## [1.0.2]

* Updated pub.dev health recommendations

## [1.0.1]

* Upgrade intl package

## [1.0.0]

* We consider the library as stable

## [0.1.9]

* Fixed small offset issue between yaxis and line/point drawing

## [0.1.8]

* Added gradient option to the area chart

## [0.1.7]

* More detailed description

## [0.1.6]

* Updated readme

## [0.1.5]

* Updated readme

## [0.1.4]

* Adding support for area chart

## [0.1.3]

* Formatting axis values dynamic depending on the tick size

## [0.1.2]

* Improve performance

## [0.1.1]

* Calculate axis width, to make the chart paint full width

## [0.1.0]

* fix drawing series with all values being the same

## [0.0.8]

* reduce margins

## [0.0.7]

* show tooltip on tap
* coverage back to 100

## [0.0.6]

* Change animations, to be really cool, now the animations draws the chart lines.
* secondary Y axis support, to support 1-2 units.

## [0.0.5]

* Restyled tooltip
* Finish testing

## [0.0.4]

* Style tooltip
* Paint unit
* Make sure tooltip paints within chart area
* Formatted dynamic datetime axisvalues

## [0.0.3]

* Added codemagic CI

## [0.0.2]

* Added screenshots
* Fixed warnings

## [0.0.1]

* Very initial release, to include the library in our project


example/lib/main.dart
import 'package:example/fake_chart_series.dart';
import 'package:fl_animated_linechart/chart/area_line_chart.dart';
import 'package:fl_animated_linechart/common/pair.dart';
import 'package:fl_animated_linechart/fl_animated_linechart.dart';
import 'package:flutter/material.dart';

void main() => runApp(MyApp());

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      debugShowCheckedModeBanner: false,
      title: 'Flutter Demo',
      theme: ThemeData(
        primarySwatch: Colors.blue,
      ),
      home: MyHomePage(title: 'fl_animated_chart demo'),
    );
  }
}

class MyHomePage extends StatefulWidget {
  MyHomePage({Key? key, this.title}) : super(key: key);

  final String? title;

  @override
  _MyHomePageState createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> with FakeChartSeries {
  int chartIndex = 0;

  @override
  Widget build(BuildContext context) {
    String? title = widget.title ?? '';
    Map<DateTime, double> line1 = createLine2();
    Map<DateTime, double> line2 = createLine2_2();
    Map<DateTime, double> line3 = yAxisUpperMaxMarkerLine();
    Map<DateTime, double> line4 = yAxisUpperMinMarkerLine();
    Map<DateTime, double> line5 = yAxisLowerMinMarkerLine();
    Map<DateTime, double> line6 = yAxisLowerMaxMarkerLine();
    Map<DateTime, double> line7 = createLine2_3();

    LineChart chart;

    if (chartIndex == 0) {
      chart = LineChart.fromDateTimeMaps([
        line1,
        line2,
      ], [
        Colors.green,
        Colors.blue,
      ], [
        'C',
        'C',
      ], tapTextFontWeight: FontWeight.w400);
    } else if (chartIndex == 1) {
      chart = LineChart.fromDateTimeMaps(
          [createLineAlmostSaveValues()], [Colors.green], ['C'],
          tapTextFontWeight: FontWeight.w400);
    } else if (chartIndex == 2) {
      chart = AreaLineChart.fromDateTimeMaps(
          [line1], [Colors.red.shade900], ['C'],
          yAxisName: "Temperature",
          gradients: [Pair(Colors.yellow.shade400, Colors.red.shade700)]);
    } else {
      chart = LineChart.fromDateTimeMaps([
        line7,
        line3,
        line4,
        line5,
        line6,
      ], [
        Colors.blue,
        Colors.red,
        Colors.yellow,
        Colors.yellow,
        Colors.red,
      ], [
        'C',
        'C',
        'C',
        'C',
        'C',
      ], tapTextFontWeight: FontWeight.w400);
      chart.lines[1].isMarkerLine = true;
      chart.lines[2].isMarkerLine = true;
      chart.lines[3].isMarkerLine = true;
      chart.lines[4].isMarkerLine = true;
    }
    final ButtonStyle flatButtonStyle = TextButton.styleFrom(
      shape: RoundedRectangleBorder(
          side: BorderSide(color: Colors.black45),
          borderRadius: BorderRadius.all(Radius.circular(3))),
    );

    return Scaffold(
      appBar: AppBar(
        title: Text(title),
      ),
      body: Container(
        child: Column(
            mainAxisSize: MainAxisSize.max,
            mainAxisAlignment: MainAxisAlignment.spaceBetween,
            crossAxisAlignment: CrossAxisAlignment.stretch,
            children: [
              Padding(
                padding: const EdgeInsets.all(8.0),
                child: Row(
                  mainAxisSize: MainAxisSize.max,
                  mainAxisAlignment: MainAxisAlignment.spaceAround,
                  children: <Widget>[
                    TextButton(
                      style: flatButtonStyle,
                      child: Text(
                        'LineChart',
                        style: TextStyle(
                            color: chartIndex == 0
                                ? Colors.black
                                : Colors.black12),
                      ),
                      onPressed: () {
                        setState(() {
                          chartIndex = 0;
                        });
                      },
                    ),
                    TextButton(
                      style: flatButtonStyle,
                      child: Text('LineChart2',
                          style: TextStyle(
                              color: chartIndex == 1
                                  ? Colors.black
                                  : Colors.black12)),
                      onPressed: () {
                        setState(() {
                          chartIndex = 1;
                        });
                      },
                    ),
                    TextButton(
                      style: flatButtonStyle,
                      child: Text('AreaChart',
                          style: TextStyle(
                              color: chartIndex == 2
                                  ? Colors.black
                                  : Colors.black12)),
                      onPressed: () {
                        setState(() {
                          chartIndex = 2;
                        });
                      },
                    ),
                    TextButton(
                      style: flatButtonStyle,
                      child: Text('MarkerLines',
                          style: TextStyle(
                              color: chartIndex == 3
                                  ? Colors.black
                                  : Colors.black12)),
                      onPressed: () {
                        setState(() {
                          chartIndex = 3;
                        });
                      },
                    ),
                  ],
                ),
              ),
              Expanded(
                  child: Padding(
                padding: const EdgeInsets.all(8.0),
                child: AnimatedLineChart(
                  chart,
                  key: UniqueKey(),
                  gridColor: Colors.black54,
                  textStyle: TextStyle(fontSize: 10, color: Colors.black54),
                  toolTipColor: Colors.white,
                  legends: chartIndex == 3
                      ? [
                          Legend(
                            title: 'Critical',
                            color: Colors.red,
                            showLeadingLine: true,
                            style: TextStyle(
                              color: Colors.black,
                            ),
                          ),
                          Legend(
                            title: 'Warning',
                            color: Colors.yellow,
                            showLeadingLine: true,
                            style: TextStyle(
                              color: Colors.black,
                            ),
                          ),
                          Legend(
                            title: 'Warning',
                            color: Colors.yellow,
                            icon: Icon(
                              Icons.report_problem_rounded,
                              size: 17,
                              color: Colors.yellow,
                            ),
                            style: TextStyle(
                              color: Colors.black,
                            ),
                          ),
                          Legend(
                            title: 'Critical',
                            color: Colors.red,
                            showLeadingLine: true,
                            style: TextStyle(
                              color: Colors.black,
                            ),
                          ),
                        ]
                      : [],
                  showMarkerLines: chartIndex == 3 ? true : false,
                  verticalMarkerColor: chartIndex == 3 ? Colors.red : null,
                  verticalMarker: [
                    DateTime.parse('2012-02-27 13:08:00'),
                    DateTime.parse('2012-02-27 13:16:00')
                  ],
                  verticalMarkerIcon: [
                    Icon(
                      Icons.cancel_rounded,
                      color: Colors.red,
                    ),
                    Icon(
                      Icons.check_circle_rounded,
                      color: Colors.green,
                    ),
                  ],
                  iconBackgroundColor: Colors.white,
                  legendsRightLandscapeMode: true,
                ), //Unique key to force animations
              )),
              SizedBox(width: 200, height: 50, child: Text('')),
            ]),
      ),
    );
  }
}
